Disk image QoL

Michaelangel007 opened the pull request:
See issue #1363 for details.

tomcw commented:
For the "applewin_boot_sector" folder, and for consistency with _prior art_...

- rename to "BootSector" and move this folder under "firmware"
- rename "c.bat" to "build.bat"
- switch from merlin32 to acme assembler? (for consistency with all the other projects under "firmware")

tomcw commented on source/Windows/WinFrame.cpp:
```diff
@@ -2093,8 +2095,309 @@ void Win32Frame::ProcessButtonClick(int button, bool bFromButtonUI /*=false*/)
+		*dst_ = g_aInterleave_DSK[ iSector ]*DSK_SECTOR_SIZE;
+	}
+
+	// With C++17 could use std::filesystem::path
```
So just use the C++17 function?
btw, typo: "retuns" -> "returns"

Michaelangel007 replied:
Fixed typo with [cb773ba](https://github.com/AppleWin/AppleWin/pull/1364/commits/cb773ba644b4c9fb947b007c1f0d3eb560a9d6f8)

Michaelangel007 replied:
AppleWin is still using C++14. Project Properties > C/C++ > Language > C++ Language Standard: **Default (ISO C++14 Standard)**

I've added support that if C++17 is used then to use that in [e57c89c](https://github.com/AppleWin/AppleWin/pull/1364/commits/e57c89c8d71d72ea951ce01289ddf9d43ffc2649)

tomcw commented on source/Windows/WinFrame.cpp:
```diff
@@ -2211,6 +2529,286 @@ void Win32Frame::ProcessDiskPopupMenu(HWND hwnd, POINT pt, const int iDrive)
+					char   *pDiskBuffer = new char[ nDiskSize ];
+					memset( pDiskBuffer, 0, nDiskSize );
+
+					// See: applewin_boot_sector
```
Rename "applewin_boot_sector" to "BootSector" ?

Michaelangel007 replied:
My thinking was that we could provide a command-line so a user could provide their own boot sector.

Let me add that and cleanup these wrongly-named `g` prefixed variables in `ProcessDiskPopupMenu()`

Michaelangel007 replied:
* Added command-line `-bootsector <file>` with [3dce821](https://github.com/AppleWin/AppleWin/pull/1364/commits/3dce8214dd8b2cf00e6ad39baf8bb175af76f88d)
* Updated path comment with [9f6b7df](https://github.com/AppleWin/AppleWin/pull/1364/commits/9f6b7df0cbf354ccb8f151cf662fff0fba37316f)

tomcw commented on source/Windows/WinFrame.cpp:
```diff
@@ -2211,6 +2529,286 @@ void Win32Frame::ProcessDiskPopupMenu(HWND hwnd, POINT pt, const int iDrive)
+					uint8_t *gpBootSector = aAppleWinBootSector;
+					if (bIsHardDisk && gbAppleWinBootSector)
+					{
+						assert( sizeof(aAppleWinBootSector) == 256 );
```
Use `static_assert`, so that it's compile-time checked.

Michaelangel007 replied:
Fixed with [d59e3a](https://github.com/AppleWin/AppleWin/pull/1364/commits/d59e3ad1e484f90c10f4d9adc0c6e41e946f2252)

tomcw commented on source/Windows/WinFrame.cpp:
```diff
@@ -2211,6 +2529,286 @@ void Win32Frame::ProcessDiskPopupMenu(HWND hwnd, POINT pt, const int iDrive)
+						0xA0,0xCC,0xC1,0xD3,0xD4,0xA0,0xD3,0xCC,0xCF,0xD4,0x00,0x00,0x00,0xD6,0xB2,0x82
+					};
+					bool     gbAppleWinBootSector = true; // TODO: allow custom boot loader via command line: -bootsector <file>
+					size_t   gnBootSectorSize = 256;
```
Use `sizeof(aAppleWinBootSector)` instead of `256`.

Michaelangel007 replied:
Fixed with commit [f7bd8e6](https://github.com/AppleWin/AppleWin/pull/1364/commits/f7bd8e62b35e9d68f55b5a337a5caeb7593fbcfc)

tomcw commented on source/Windows/WinFrame.cpp:
```diff
@@ -2211,6 +2529,286 @@ void Win32Frame::ProcessDiskPopupMenu(HWND hwnd, POINT pt, const int iDrive)
+						char Message[ 256 ];
+						if (nDiskSize < nMinDiskSize)
+						{
+							sprintf( Message, "ERROR: Disk Image Size (%zu bytes) < minimum DOS 3.3 image size (%zu bytes)", nDiskSize, nMinDiskSize );
```
Prefer `sprintf_s()` instead of `sprintf()` here and again just below.

Michaelangel007 replied:
Thanks for catching this!  Fixed 4 instances with [d6421a8](https://github.com/AppleWin/AppleWin/pull/1364/commits/d6421a81856b006d5b88dbc6e1f814cc12f7d8e8)

tomcw replied:
Hmm, this doesn't look right, as the 2nd parameter should be the buffer's size:
```
int sprintf_s(
   char *buffer,
   size_t sizeOfBuffer,
   const char *format,
   ...
);
```

Michaelangel007 replied:
sprintf_s() is overloaded with 2 versions.

I did a quick search through the code-bases and there are other usages in WinFrame.cpp

```c
Code	File	Line	Column
				if (g_nSector[slot][0] >= 0) sprintf_s(sector, "%02X", g_nSector[slot][0]);	C:\Dev\a2_projects\AppleWin\WIP_disk_qol\WinFrame.cpp	1693	34
				if (g_nSector[slot][0] >= 0) sprintf_s(sector, "%02d", g_nSector[slot][0]);	C:\Dev\a2_projects\AppleWin\WIP_disk_qol\WinFrame.cpp	1698	34
				if (g_nSector[slot][1] >= 0) sprintf_s(sector, "%02X", g_nSector[slot][1]);	C:\Dev\a2_projects\AppleWin\WIP_disk_qol\WinFrame.cpp	1704	34
				if (g_nSector[slot][1] >= 0) sprintf_s(sector, "%02d", g_nSector[slot][1]);	C:\Dev\a2_projects\AppleWin\WIP_disk_qol\WinFrame.cpp	1709	34
```

tomcw replied:
I spotted (and forgot) in the AppleWin [coding standards](https://github.com/AppleWin/AppleWin/blob/master/docs/CodingConventions.txt), we agreed not to use `sprintf()` and similar, and instead use `Strformat()`.

Michaelangel007 replied:
OK, I'll switch `StrFormat()` tomorrow.

tomcw commented on applewin_boot_sector/bootsector:
This binary file doesn't need to be included.

Michaelangel007 commented on source/ProDOS_FileSystem.h:
```diff
@@ -0,0 +1,385 @@
+	// Due to packing/aignment this may NOT be size identical
+	// But it is in the correct order
+	// See: ProDOS_GetVolumeHeader;
+	struct ProDOS_VolumeHeader_t
```
No, because all the fields are read/written individually. See `ProDOS_SetVolumeHeader()` and `ProDOS_GetFileHeader()`

Michaelangel007 commented on source/Windows/WinFrame.cpp:
```diff
@@ -2211,6 +2529,286 @@ void Win32Frame::ProcessDiskPopupMenu(HWND hwnd, POINT pt, const int iDrive)
+					fclose( hFile );
+				}
+				else
+					MessageBox( hwnd, TEXT("ERROR: Couldn't open new disk image."), pTitle, MB_OK );
```
Fixed with [ad05fda](https://github.com/AppleWin/AppleWin/pull/1364/commits/ad05fda946a65631be36838b3f0a20af19a3a375)

Michaelangel007 commented:
Where is the official repository for ACME?

* https://github.com/martinpiper/ACME.git ?

Michaelangel007 commented:
[x] rename `applewin_boot_sector` to `Firmware/BootSector`
[x] rename `c.bat` to `build.bat`
[x] Add Acme `bootsector.a`
[x] Add `Build_Merlin.bat`
[x] Add `Compare.bat` to verify Acme is byte identical to Merlin32 source

tomcw commented:
Delete the `Apple2.bin` file, as this isn't used in the build of AppleWin?

Also delete the Merlin asm source file (`bootsector.s`) ?
Which means you can also delete `Build_Merlin.bat` and `Compare.bat`.

Michaelangel007 commented:
> Delete the `Apple2.bin` file, as this isn't used in the build of AppleWin?

It is used in the Acme assembly source of the `bootsector.a` due to it having half-assed text processing.

Michaelangel007 commented:
> Delete the `Apple2.bin` file, as this isn't used in the build of AppleWin?
> 
> Also delete the Merlin asm source file (`bootsector.s`) ? Which means you can also delete `Build_Merlin.bat` and `Compare.bat`.

I'd rather keep the original Merlin source since that is my _greatly preferred_ assembler.

tomcw commented:
>I'd rather keep the original Merlin source since that is my _greatly preferred_ assembler.

Fair enough.

I still think it's better to include the binary directly as a resource, instead of inlining a C array.

Michaelangel007 commented:
**Update:** Sorry tomcw for editing your message. I thought I was editing my reply.

> I still think it's better to include the binary directly as a resource, instead of inlining a C array.

Agreed. I wanted a reference version working -- since I was going through a few iterations of the custom boot sector -- but now that v2 is stable I moved the boot sector into a native resource.

* [x] Removed inlined C array. Using `firmware/BootSector/bootsector.bin` with [fcd3f6d](https://github.com/AppleWin/AppleWin/pull/1364/commits/fcd3f6d6d804afa7ddbef8c4eafcce34d6b81ebe)

tomcw commented:
Great! So I think that you can remove `Apple2.bin` and `bin_to_c.cpp` ?

Michaelangel007 commented:
> Great! So I think that you can remove `Apple2.bin` and `bin_to_c.cpp` ?

* `bin_to_c.cpp` can be removed.

* `Apple2.bin` is the **Text Mapping file used by the ACME assembler** for when assembly source has text strings such as: `!text  "foo"`.  If you hexdump this file you'll see the high-bit has been set for all the ASCII chars.  **It should remain so people can properly assemble `bootsector.a`**

Michaelangel007 commented:
@tomcw I've updated a few help documents.

tomcw commented on help/ddi-sizes.html:
```diff
@@ -0,0 +1,31 @@
+<ul>
+  <li>A standard 5.25" floppy disk normally has 35 tracks. This disk can hold 140 KB.</li>
+
+  <li>A extended 5.25" floppy disk may have up to 40 tracks. This disk can hold 160 KB.</li>
```
typo: "An extended..."

Michaelangel007 replied:
Thanks!  Fixed with [d4e0906](https://github.com/AppleWin/AppleWin/pull/1364/commits/d4e090674f34e4508f31e9f313d5a554bbf34305)

Michaelangel007 commented:
After talking with Rikkles I'm going to make another pass on the Creating Disk Images since the flow is a little disjointed. There are really 5 ways of creating disks and the docs should reflect that.

1. Empty data disk
2. DOS 3.3 disk
3. ProDOS 8 disk
4. AppleWin Boot Sector disk
5. Custom Boot Sector
6. Legacy new filename

hasseily commented:
Should the ProDOS image have BASIC?

hasseily commented:
I just posted a small merge request, you have two dots for the image filename extension.

Michaelangel007 commented:
> I just posted a small merge request, you have two dots for the image filename extension.

Thanks, fixed!

Michaelangel007 commented:
> Should the ProDOS image have BASIC?

That's a _really_ good question.

Initially, I wasn't planning on it, due to three reasons ...

* I wanted an option to quickly format "data disks"

* https://github.com/AppleWin/AppleWin/issues/1365 including a new ProDOS master disk (which probably necessitates including new instructions to copy BASIC.SYSTEM over to a new disk.)

* Didn't want to add more ProDOS code for `Util_ProDOS_AddBasic()`

... BUT I'm open to suggestions.

I mean if we've gone down the whole "make a new disk bootable & usable" then it probably _should_ include `BASCIC.SYSTEM`. 

Let me see how much more code I need for  `UIil_ProDOS_CopyBasic()` -> `Util_ProDOS_AddFile()`.

Michaelangel007 commented:
> Should the ProDOS image have BASIC?

This was actually pretty trivial to add!  This is now in.  One less reason to need to use an external utility (or Copy ][+ is always nice.)

![disk_menu_normal_v5](https://github.com/user-attachments/assets/45bdba20-e8d3-4e19-ad7d-7f1494d2bf30)

sicklittlemonkey commented:
I think you should replace all this with a wizard ...

Just kidding! Nice work!

Michaelangel007 commented:
> I think you should replace all this with a wizard ...

... soon as I receive your PR. ;-)

> Just kidding! Nice work!

Thanks!  I'm _finally_ happy with how this turned out after having some time for this to "bake".  The checkmarks to allow a user to select what they want (or don't) on a ProDOS should take care of 99% of users.

tomcw commented:
@Michaelangel007 - is this PR ready to merge? (Or are you still working on it... which is fine)

Michaelangel007 commented:
Yes, still working on this.

* [ ] Track down off-by-one block when `PRODOS` is copied onto a disk
* [ ] Update Help
* [ ] Switch to `StrFormat`

tomcw commented:
@Michaelangel007 - I want to do a new release next weekend (8th March). Any chance this PR can be wrapped up by then?

tomcw commented:
btw, you can remove "firmware/BootSector/apple2.bin", as this file is already included in the ACME distro.

You'll need to create an ENV VAR, eg.
`ACME=C:/.../acme/ACME_Lib`
(Yes - it needs forward slashes!)

And then change "firmware/BootSector/bootsector.a" (line 3):
`!ct "apple2.bin"`
to
`!ct <apple ii/convtab.bin>`

Michaelangel007 commented:
**Note:** [convtab.bin](https://github.com/visrealm/acme/blob/main/ACME_Lib/apple%20ii/convtab.bin) can be found in `acme/ACME_Lib/apple ii/convtab.bin` -> `apple2.bin`

Hexdump:
```
00000000: 80 81 82 83 84 85 86 87  88 89 8A 8B 8C 8D 8E 8F  ................
00000010: 90 91 92 93 94 95 96 97  98 99 9A 9B 9C 9D 9E 9F  ................
00000020: A0 A1 A2 A3 A4 A5 A6 A7  A8 A9 AA AB AC AD AE AF  ................
00000030: B0 B1 B2 B3 B4 B5 B6 B7  B8 B9 BA BB BC BD BE BF  ................
00000040: C0 C1 C2 C3 C4 C5 C6 C7  C8 C9 CA CB CC CD CE CF  ................
00000050: D0 D1 D2 D3 D4 D5 D6 D7  D8 D9 DA DB DC DD DE DF  ................
00000060: E0 E1 E2 E3 E4 E5 E6 E7  E8 E9 EA EB EC ED EE EF  ................
00000070: F0 F1 F2 F3 F4 F5 F6 F7  F8 F9 FA FB FC FD FE FF  ................
00000080: 00 01 02 03 04 05 06 07  08 09 0A 0B 0C 0D 0E 0F  ................
00000090: 10 11 12 13 14 15 16 17  18 19 1A 1B 1C 1D 1E 1F  ................
000000A0: 20 21 22 23 24 25 26 27  28 29 2A 2B 2C 2D 2E 2F   !"#$%&'()*+,-./
000000B0: 30 31 32 33 34 35 36 37  38 39 3A 3B 3C 3D 3E 3F  0123456789:;<=>?
000000C0: 40 41 42 43 44 45 46 47  48 49 4A 4B 4C 4D 4E 4F  @ABCDEFGHIJKLMNO
000000D0: 50 51 52 53 54 55 56 57  58 59 5A 5B 5C 5D 5E 5F  PQRSTUVWXYZ[\]^_
000000E0: 60 61 62 63 64 65 66 67  68 69 6A 6B 6C 6D 6E 6F  `abcdefghijklmno
000000F0: 70 71 72 73 74 75 76 77  78 79 7A 7B 7C 7D 7E 7F  pqrstuvwxyz{|}~.
```

tomcw commented on help/ddi-create.html:
```diff
@@ -11,27 +11,59 @@
+<h5>Manually putting DOS 3.3 on a blank disk</h5>
 
-<p><b>NOTE:</b> This disk will NOT be <i>bootable</i>. You will need to <i>format</i> the disk (keep reading) to put a File System on it and (optionally) copy a Disk Operating System (DOS 3.3, ProDOS, etc.) onto it.  This will make the disk bootable.</p>
+<p><b>NOTE:</b> When you This disk will NOT be <i>bootable</i>. You will need to <i>format</i> the disk (keep reading) to put a File System on it and (optionally) copy a Disk Operating System (DOS 3.3, ProDOS, etc.) onto it.  This will make the disk bootable.</p>
```
Typo here: "When you"

tomcw commented on help/ddi-create.html:
```diff
@@ -80,27 +112,39 @@ <h3>Classic Disk Creation</h3>
+  <li>.DO -- DOS 3.3 sector order</li>
+  <li>.PO -- ProDOS sector order</li>
+  <li>.HDV -- ProDOS (hard drive) volume in ProDOS sector order</li>
+  <li>.NIB -- and older raw nibble image with 256 bytes/sector, and</li>
```
Probably more correct to say "0x1A00 (6656) nibbles per track" (as .NIB can theoretically represent weird formats that don't have 256 bytes/sector).

Michaelangel007 commented:
@tomcw I've made another pass through `help/ddi-create.html` -- fixed spelling, grammar, flow.  I've split it into two files now since it was getting kind of cluttered:

* `help/ddi-create.html` -- Creating Disk Images
* `help/ddi-advanced.html` -- Advanced Formatting Options

I'm pretty happy with the former so if you want to make another pass through that would be appreciated.

I'm currently plugging away on getting `help/ddi-advanced.html` into a shippable state.

tomcw commented on resource/resource.h:
```diff
@@ -132,19 +139,37 @@
-#define _APS_NEXT_COMMAND_VALUE         40012
-#define _APS_NEXT_CONTROL_VALUE         1083
+#define _APS_NEXT_RESOURCE_VALUE        167
+#define _APS_NEXT_COMMAND_VALUE         40022
```
FWIW, this should be 40026 (not 40022).

Michaelangel007 replied:
I thought I fixed that.  Oh well.  Thanks for the catch!

tomcw commented on help/ddi-create.html:
```diff
@@ -62,10 +120,26 @@ <h2><font color="#008000">Creating
+  <li>.DO -- DOS 3.3 sector order,</li>
+  <li>.PO -- ProDOS sector order,</li>
+  <li>.HDV -- ProDOS (hard drive) volume in ProDOS sector order,</li>
+  <li>.NIB -- and older raw nibble image that supports up to 6656 (0x1A00) nibbles/tracks, and</li>
```
(Pedantic) maybe remove the "and" at _both_ the start & end of this NIB explanation.

Michaelangel007 replied:
Agreed. Thanks for catching that.

tomcw commented:
> I'm pretty happy with the former so if you want to make another pass through that would be appreciated.

Just one (pedantic) comment - see above.

> I've split it into two files now since it was getting kind of cluttered

Good idea - and I agree, this works better.

Michaelangel007 commented on help/img/disk_select_bootsector_prodos.png:
Ah, those images were created before we settled on the new DOS  3.3 System Master file.

Thanks for catching those.

tomcw commented on help/ddi-create.html:
```diff
@@ -11,46 +11,104 @@
+    </ul>
+</p>
+
+<p>Newer programs tend use ProDOS while older programs tend to use DOS 3.3.</p></li>
```
Typo: "Newer programs tend *to* use..."

Michaelangel007 commented on help/ddi-create.html:
```diff
@@ -11,46 +11,104 @@
+<ol>
+  <li><strong>Right-Click</strong> on the Drive 1 <img src="img/tb-drv1.png" /> toolbar button (or press <span style="font-family: Courier New,Courier,monospace;">Ctrl-F3</span>) to open the drive pop-up menu.<br>
+
+  <br><img src="img/disk_new_140kb_disk.png" /></li>
```
Oooh, it does. I guess Win10 defaults to a gray border to the black really makes it "pop".

tomcw commented:
@Michaelangel007 - I've updated the .hhc (TOC) file to include your new ddi-advanced.html page.

Michaelangel007 commented:
> I've updated the .hhc (TOC) file to include your new ddi-advanced.html page.

I was going to do that after I was done updating it but thanks!

I've made some progress on the advanced page today.  I will keep working on it tomorrow and try to get it wrapped up ASAP.

You may have noticed I've added a round HTML5 border around "keys".

Q. **Do we want to do this for the rest of the HTML help?**  (I haven't had a chance to look at the `.chm` to see how good/bad this looks, yet.)

tomcw commented:
>You may have noticed I've added a round HTML5 border around "keys".

Yes - looks good!

>Q. Do we want to do this for the rest of the HTML help?

Let's wrap up this ddi-advanced.html first, then review this later.
btw, after building the AppleWin.chm, the border isn't rounded, but square (when viewed from the chm).

Michaelangel007 commented:
@tomcw I made a pass through the _Advanced Formatting Options_ (`help/ddi-advanced.html`)  I _think_ it is ready if you wouldn't mind taking a look.

tomcw commented:
Looks and reads great!

Some of the images could do with scaling down to a smaller size, eg: disk_boot_data_disk.png, disk_select.png_blank_dos.png and disk_select_blank_prodos.png.

And those last 2 still have the old MASTER.DSK shown in them.

Otherwise it's ready.

Michaelangel007 commented:
@tomcw I took a look at re-scaling the images a few days ago via via HTML5:

```JavaScript
<img style="width: 50%; height: 50%;" />
```
but that look hideous so I backed out of it. (I didn't realize it at the time but I was not understanding the correct scaling usage.)

I also tried using a transform, something along these lines:
```JavaScript
transform: scale(0.5);
```
but that kept the original border size (!) but shrunk the image down by 50% which looked funny in a shrunken-head way.

Today I was going to manually the images but after doing a test ...

* 4K, native 100%
![test_4K_native_one_one](https://github.com/user-attachments/assets/e106bb20-2a5f-4ead-9c37-f65c349f9030)

* 4K, 150%, manually scaled down 66% to match the original 4K 100% UI dimensions:
![test_4k_150percent_downscale_66_percent](https://github.com/user-attachments/assets/7ae10eae-cb64-44a0-8b04-8b1497f61f91)

... I see that the 66% manual downscaling looks fugly (blurry as hell) and I'm not a fan of re-taking screenshots since that was going to take a bit of work to re-capture everything ... so I took another look at scaling via HTML5 but this time using:

```JavaScript
<img style="width: 33%; height: auto;" />
```
Turns out this is scaling the image relative to the **page width** and not the original image dimensions.

I've settled on using 33% for the menus and 50% for full screen images.  I looked at it on both in landscape and portrait mode and it seems to look good so I'm going with this.

I'm willing to bet there is _probably_ a combination of `max-width`, `object-fit`, `object-position`, or some other CSS I could use but this is good enough for now.
 
 I sometimes wonder if us developers _don't_ add features just so we don't have to update the help / documentation ... ;-)

tomcw commented:
>I've settled on using 33% for the menus and 50% for full screen images. I looked at it on both in landscape and portrait mode and it seems to look good so I'm going with this.

Look good!

>I sometimes wonder if us developers don't add features just so we don't have to update the help / documentation ... ;-)

:-)

So I think we are done?

tomcw commented:
Merged it! Thanks @Michaelangel007 for all this great work!
